File: survivor-picker/strategy/entry_a_value.py
"""Weekly pick strategy for Entry A.

Combines this week's win probabilities (``models.win_prob``) with each
team's future value (``models.future_value``) into a single score:

    score = win_pct * (1 - future_value_penalty)

``future_value_penalty`` is how much we discount a team for this week
because a materially better matchup is projected in the next few weeks --
so a big favorite with an even bigger mismatch coming up in a week or two
can rank behind a smaller, "use it now" favorite. The penalty is capped
(see ``MAX_FUTURE_VALUE_PENALTY``) so a distant hypothetical matchup can
never fully override a strong matchup in hand.

This module only reasons about Entry A -- ``used_teams_a.json`` via
``state/entries_store.py``. It produces a ranked list and a top pick with
plain-English reasoning; it never submits anything.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from data.models import Game
from models.future_value import DEFAULT_DECAY_RATE, DEFAULT_LOOKAHEAD_WEEKS, compute_future_value
from models.win_prob import TeamWeekWinProbability, resolve_team_win_probability
from state.entries_store import load_used_teams_for_entry

ENTRY_NAME = "Entry A"

# A future_value of this many win-probability points (or more) maps to the
# full penalty cap below. E.g. with the defaults, a future matchup that's
# 40 points better (after decay) than this week's caps the discount at 35%.
FUTURE_VALUE_PENALTY_SCALE = 40.0
MAX_FUTURE_VALUE_PENALTY = 0.35


@dataclass
class RankedPick:
    team_abbreviation: str
    opponent_abbreviation: Optional[str]
    win_pct: Optional[float]  # 0-100, may be spread-estimated
    win_pct_source: str  # "api" | "spread_estimate" | "unknown"
    spread_detail: Optional[str]
    future_value: Optional[float]  # points of upside from holding this team back, may be negative/None
    future_value_penalty: float  # 0 to MAX_FUTURE_VALUE_PENALTY, fraction actually applied
    score: Optional[float]  # win_pct * (1 - future_value_penalty), None if win_pct is unknown


@dataclass
class EntryARecommendation:
    week: int
    pick: Optional[RankedPick]
    reasoning: str
    alternatives: List[RankedPick] = field(default_factory=list)


def _future_value_to_penalty(future_value: Optional[float]) -> float:
    """Map a future_value (win-pct-point delta) to a capped 0-1 penalty.

    Only a *positive* future_value (a better matchup projected later)
    produces a penalty -- there's never a bonus for a team with nothing
    better coming up.
    """
    if future_value is None or future_value <= 0:
        return 0.0
    return min(MAX_FUTURE_VALUE_PENALTY, future_value / FUTURE_VALUE_PENALTY_SCALE)


def rank_available_teams(
    current_week_games: List[Game],
    win_prob_table: Dict[Tuple[str, int], TeamWeekWinProbability],
    used_teams: List[str],
    current_week: int,
    lookahead_weeks: int = DEFAULT_LOOKAHEAD_WEEKS,
    decay_rate: float = DEFAULT_DECAY_RATE,
) -> List[RankedPick]:
    """Rank this week's not-yet-used teams by ``win_pct * (1 - future_value_penalty)``.

    ``current_week_games`` supplies this week's win probability and spread
    per team; ``win_prob_table`` (spanning the season, built by
    ``models.win_prob.build_win_probability_table``) supplies the future
    matchups used to compute each team's future value. A team with no
    ``win_pct`` at all (score ``None``) sorts last -- there's no basis to
    recommend it over a team we can actually score.
    """
    ranked: List[RankedPick] = []

    for game in current_week_games:
        if game.state and game.state != "pre":
            continue  # already started/finished isn't a legitimate pick for this week
        spread_detail = game.odds.details if game.odds else None

        for team, opponent, is_home in ((game.home, game.away, True), (game.away, game.home, False)):
            if not team.abbreviation or team.abbreviation in used_teams:
                continue

            resolved = resolve_team_win_probability(game, is_home)
            # This week's resolved win_pct is the baseline -- not a lookup
            # into win_prob_table for the current week, which may not
            # include it (or may be a stale cached copy of it).
            remaining_schedule = [
                entry
                for (abbrev, wk), entry in win_prob_table.items()
                if abbrev == team.abbreviation and wk > current_week
            ]
            future = compute_future_value(
                team.abbreviation,
                current_week,
                resolved.win_pct,
                remaining_schedule,
                lookahead_weeks=lookahead_weeks,
                decay_rate=decay_rate,
            )
            penalty = _future_value_to_penalty(future.future_value)
            score = resolved.win_pct * (1 - penalty) if resolved.win_pct is not None else None

            ranked.append(
                RankedPick(
                    team_abbreviation=team.abbreviation,
                    opponent_abbreviation=opponent.abbreviation,
                    win_pct=resolved.win_pct,
                    win_pct_source=resolved.source,
                    spread_detail=spread_detail,
                    future_value=future.future_value,
                    future_value_penalty=penalty,
                    score=score,
                )
            )

    ranked.sort(key=lambda p: (p.score is None, -(p.score or 0)))
    return ranked


def _describe_pick(pick: RankedPick) -> str:
    win_pct = f"{pick.win_pct:.1f}%" if pick.win_pct is not None else "unknown"
    basis = " (estimated from spread)" if pick.win_pct_source == "spread_estimate" else ""
    spread = f", spread {pick.spread_detail}" if pick.spread_detail else ""
    return f"{pick.team_abbreviation} vs {pick.opponent_abbreviation or '?'} -- {win_pct} win prob{basis}{spread}"


def _build_reasoning(top: RankedPick, alternatives: List[RankedPick]) -> str:
    parts = [f"Top pick: {_describe_pick(top)}."]

    if top.score is None:
        parts.append("No win probability data was available for this pick; it was chosen by default ordering.")
    elif top.future_value_penalty > 0:
        parts.append(
            f"A future-value penalty of {top.future_value_penalty:.0%} was applied "
            f"(a projected future matchup is about {top.future_value:.1f} points better after decay), "
            f"but {top.team_abbreviation} still scored highest at {top.score:.1f}."
        )
    else:
        parts.append(
            f"No upcoming matchup is projected to beat this week's, so there's little value in holding "
            f"{top.team_abbreviation} back -- it scored {top.score:.1f} with no penalty applied."
        )

    if alternatives:
        runner_up = alternatives[0]
        runner_score = f"{runner_up.score:.1f}" if runner_up.score is not None else "unknown"
        top_score = f"{top.score:.1f}" if top.score is not None else "unknown"
        parts.append(f"Next best was {_describe_pick(runner_up)}, scoring {runner_score} vs {top_score}.")

    return " ".join(parts)


def recommend(
    current_week_games: List[Game],
    win_prob_table: Dict[Tuple[str, int], TeamWeekWinProbability],
    current_week: int,
    lookahead_weeks: int = DEFAULT_LOOKAHEAD_WEEKS,
    decay_rate: float = DEFAULT_DECAY_RATE,
    used_teams: Optional[List[str]] = None,
) -> EntryARecommendation:
    """Top pick recommendation for Entry A this week, with reasoning.

    ``used_teams`` defaults to loading ``state/used_teams_a.json``; pass it
    explicitly to keep this function pure for testing.
    """
    if used_teams is None:
        used_teams = load_used_teams_for_entry(ENTRY_NAME)

    ranked = rank_available_teams(
        current_week_games, win_prob_table, used_teams, current_week, lookahead_weeks, decay_rate
    )

    if not ranked:
        return EntryARecommendation(
            week=current_week,
            pick=None,
            reasoning="No eligible teams available this week (all used, or no game data).",
            alternatives=[],
        )

    top, alternatives = ranked[0], ranked[1:]
    return EntryARecommendation(
        week=current_week,
        pick=top,
        reasoning=_build_reasoning(top, alternatives),
        alternatives=alternatives,
    )

File: deadpool/src/engine/strategies/entry-a-value.js
/**
 * Win probability, discounted by what the team is worth later.
 *
 * Port of `strategy/entry_a_value.py`, unchanged.
 *
 *     score = winPct * (1 - futureValuePenalty)
 *
 * The penalty is how much a team is marked down this week because a materially
 * better matchup is projected in the next few weeks — so a big favourite with
 * an even bigger mismatch coming in a week or two can rank behind a smaller,
 * use-it-now favourite. It is capped, so a distant hypothetical can never
 * fully override a strong matchup in hand.
 *
 * ── The part that has never run ─────────────────────────────────────────
 *
 * This strategy is only itself when it is given a schedule that extends past
 * the current week. In the Python CLI it never was: `main.py` does not import
 * this module at all, and even wired up it would have been fed a table built
 * from one week of games, leaving `remainingSchedule` empty, `futureValue`
 * null and the penalty flat zero — at which point it is `recommender.js` with
 * extra steps. The tests pass because they hand it a multi-week table by hand.
 *
 * The engine gets the real thing because ctx.schedule comes from the edge
 * Function's /api/season route. Nothing here changed to make that true.
 */

import { buildOptions, notUsed, byScoreDesc } from '../constraints.js';
import { computeFutureValue, DEFAULT_DECAY_RATE, DEFAULT_LOOKAHEAD_WEEKS } from '../future-value.js';
import { basisPhrase, remainingScheduleFor } from '../win-prob.js';
import { f1, pct0 } from '../fmt.js';
import { findConflicts } from './recommender.js';

export const ID = 'value';

// A futureValue of this many win-probability points (or more) maps to the full
// cap below: with the defaults, a future matchup 40 points better after decay
// caps the discount at 35%.
export const FUTURE_VALUE_PENALTY_SCALE = 40.0;
export const MAX_FUTURE_VALUE_PENALTY = 0.35;

/**
 * A futureValue in win-percentage points → a capped 0–1 penalty.
 *
 * Only a positive futureValue produces a penalty. There is deliberately no
 * bonus for a team with nothing better coming: the question this asks is
 * "should I wait", and "no" is not a reason to prefer them.
 */
export function futureValueToPenalty(futureValue) {
  if (futureValue === null || futureValue === undefined || futureValue <= 0) return 0.0;
  return Math.min(MAX_FUTURE_VALUE_PENALTY, futureValue / FUTURE_VALUE_PENALTY_SCALE);
}

/**
 * Rank this week's not-yet-used teams by the discounted score.
 *
 * `winProbTable` spans the season and supplies the future matchups; this
 * week's own probability comes from the game in hand rather than from a
 * lookup, which may be a stale cached copy of it.
 */
export function rankAvailableTeams(
  games,
  winProbTable,
  usedTeams,
  currentWeek,
  lookaheadWeeks = DEFAULT_LOOKAHEAD_WEEKS,
  decayRate = DEFAULT_DECAY_RATE,
) {
  const ranked = notUsed(buildOptions(games), usedTeams).map((o) => {
    const remaining = remainingScheduleFor(winProbTable, o.teamAbbreviation, currentWeek);
    const future = computeFutureValue(
      o.teamAbbreviation, currentWeek, o.winPct, remaining, lookaheadWeeks, decayRate,
    );
    const penalty = futureValueToPenalty(future.futureValue);
    return {
      teamAbbreviation: o.teamAbbreviation,
      teamName: o.teamName,
      opponentAbbreviation: o.opponentAbbreviation,
      eventId: o.eventId,
      startDate: o.startDate,
      winPct: o.winPct,
      winPctSource: o.winPctSource,
      winPctIsEstimated: o.winPctSource === 'spread_estimate',
      spreadDetail: o.spreadDetail,
      futureValue: future.futureValue,
      futureValuePenalty: penalty,
      bestFutureWeek: future.bestFutureWeek,
      bestFutureWinPct: future.bestFutureWinPct,
      score: o.winPct === null || o.winPct === undefined ? null : o.winPct * (1 - penalty),
    };
  });
  return ranked.sort(byScoreDesc);
}

export function describePick(pick) {
  const winPct = pick.winPct === null || pick.winPct === undefined ? 'unknown' : `${f1(pick.winPct)}%`;
  const basis = basisPhrase(pick.winPctSource);
  const spread = pick.spreadDetail ? `, spread ${pick.spreadDetail}` : '';
  return `${pick.teamAbbreviation} vs ${pick.opponentAbbreviation || '?'} -- ${winPct} win prob${basis}${spread}`;
}

export function buildReasoning(top, alternatives) {
  const parts = [`Top pick: ${describePick(top)}.`];

  if (top.score === null || top.score === undefined) {
    parts.push('No win probability data was available for this pick; it was chosen by default ordering.');
  } else if (top.futureValuePenalty > 0) {
    parts.push(
      `A future-value penalty of ${pct0(top.futureValuePenalty)} was applied `
      + `(a projected future matchup is about ${f1(top.futureValue)} points better after decay), `
      + `but ${top.teamAbbreviation} still scored highest at ${f1(top.score)}.`,
    );
  } else {
    parts.push(
      `No upcoming matchup is projected to beat this week's, so there's little value in holding `
      + `${top.teamAbbreviation} back -- it scored ${f1(top.score)} with no penalty applied.`,
    );
  }

  if (alternatives.length) {
    const runnerUp = alternatives[0];
    const runnerScore = runnerUp.score === null || runnerUp.score === undefined ? 'unknown' : f1(runnerUp.score);
    const topScore = top.score === null || top.score === undefined ? 'unknown' : f1(top.score);
    parts.push(`Next best was ${describePick(runnerUp)}, scoring ${runnerScore} vs ${topScore}.`);
  }

  return parts.join(' ');
}

export function recommend(games, winProbTable, currentWeek, usedTeams, lookaheadWeeks = DEFAULT_LOOKAHEAD_WEEKS, decayRate = DEFAULT_DECAY_RATE) {
  const ranked = rankAvailableTeams(games, winProbTable, usedTeams, currentWeek, lookaheadWeeks, decayRate);

  if (!ranked.length) {
    return {
      week: currentWeek,
      pick: null,
      reasoning: 'No eligible teams available this week (all used, or no game data).',
      alternatives: [],
    };
  }

  const [top, ...alternatives] = ranked;
  return { week: currentWeek, pick: top, reasoning: buildReasoning(top, alternatives), alternatives };
}

/* ------------------------------------------------ the registry contract -- */

export default {
  id: ID,
  name: 'Value, with lookahead',
  blurb: 'Marks a team down when a better matchup for them is projected in the next few weeks, so a strong team is not spent early.',
  entries: 'single',
  params: [
    { key: 'lookaheadWeeks', label: 'Look ahead', type: 'int', default: DEFAULT_LOOKAHEAD_WEEKS, min: 1, max: 12, unit: 'weeks', help: 'How far forward a better matchup is worth waiting for.' },
    { key: 'decayRate', label: 'Weekly discount', type: 'float', default: DEFAULT_DECAY_RATE, min: 0.5, max: 1, step: 0.01, help: 'How much less a matchup counts per week of distance. 1.00 trusts a week-6 projection as much as next Sunday.' },
  ],

  run(ctx) {
    const lookahead = ctx.params.lookaheadWeeks ?? DEFAULT_LOOKAHEAD_WEEKS;
    const decay = ctx.params.decayRate ?? DEFAULT_DECAY_RATE;
    const perEntry = {};
    const picks = [];

    for (const entry of ctx.entries) {
      const r = recommend(ctx.games, ctx.schedule, ctx.week, ctx.usedTeams[entry.id] ?? [], lookahead, decay);
      perEntry[entry.id] = r.pick ? [r.pick, ...r.alternatives] : [];
      picks.push({
        entry: entry.id,
        candidate: r.pick,
        reasoning: r.reasoning,
        factors: r.pick ? factorsFor(r.pick, r.alternatives[0]) : [],
      });
    }

    const warnings = [];
    if (ctx.scheduleWeeks <= 1) {
      warnings.push({ level: 'warn', text: 'Only this week is loaded, so nothing is projected ahead and the lookahead is doing nothing. This ranks identically to win probability until the season schedule is available.' });
    }
    // Two entries reasoned about one at a time land on the same team, because
    // this is deterministic and they start the season with the same inventory.
    // The backtester measures it at 100% of weeks until one of them dies --
    // two entries in name only, and the one holding that cannot hedge
    // anything. `recommender.js` has always warned about this and the two
    // per-entry strategies beside it did not; sequence-dp.js now does, and its
    // comment named this file as the remaining gap.
    const conflict = findConflicts(perEntry);
    if (conflict) {
      warnings.push({
        level: 'warn',
        text: `Both entries' top pick is ${conflict}. One result would eliminate both — take the runner-up for one of them.`,
      });
    }

    return {
      strategyId: ID,
      picks,
      candidates: perEntry,
      considered: Object.values(perEntry).reduce((n, c) => n + c.length, 0),
      warnings,
    };
  },
};

/** The prose above, as something the interface can lay out. */
function factorsFor(pick, runnerUp) {
  const f = [
    {
      label: 'Win probability',
      value: pick.winPct === null ? null : `${f1(pick.winPct)}%`,
      weight: 1,
      note: pick.winPctIsEstimated ? 'Estimated from the spread — ESPN has not published a probability yet.' : 'From ESPN.',
    },
  ];
  if (pick.futureValuePenalty > 0) {
    f.push({
      label: 'Held-back penalty',
      value: `−${pct0(pick.futureValuePenalty)}`,
      weight: -1,
      note: `Week ${pick.bestFutureWeek} looks about ${f1(pick.futureValue)} points better after decay.`,
    });
  } else {
    f.push({ label: 'Held-back penalty', value: 'none', weight: 0, note: 'Nothing better is projected for them inside the window.' });
  }
  f.push({ label: 'Score', value: pick.score === null ? null : f1(pick.score), weight: 1 });
  if (runnerUp) {
    f.push({
      label: 'Beat',
      value: runnerUp.teamAbbreviation,
      weight: 0,
      note: runnerUp.score === null ? 'The runner-up could not be scored.' : `by ${f1(pick.score - runnerUp.score)} points.`,
    });
  }
  return f;
}
